### ISSUE TITLE
More information for NeomakeFinished

### ISSUE THREAD

**Issue by teto** [neomake/neomake#767]
I would like to launch a popup when the maker finishes with an error (exit value != 0). I have the code to trigger the notification but I wonder how to retrieve the exit value ?
Right now I have :
```

function! OnNeomakeFinished()
  echo g:neomake_hook_context
endfunction

augroup my_neomake
    au!
    autocmd User NeomakeFinished call OnNeomakeFinished()
augroup END

```
but g:neomake_hook_context hardly displays anything. I don't know much about vim autcmd but wouldn't that be cleaner to pass arguments there rather than using a global variable ?
Nb: I joined #neomake@freenode

**Comment by blueyed** [comment]
AFAIK it's not possible to pass arguments to autocommands, and you typically have `v:` like variables there then (which we cannot use).

`NeomakeFinished` could be enhanced, but it is triggered when all makers have finished.

Do you want `NeomakeMakerExited` instead?  Or rather have information about all jobs in `NeomakeFinished`, e.g. a list of `jobinfo` objects?

Related: https://github.com/neomake/neomake/issues/766

**Comment by jwmann** [comment]
I'm also a bit confused with these variables.

According to the docs these variables are set
- `g:neomake_hook_context.file_mode`
- `g:neomake_hook_context.make_id`
- `g:neomake_hook_context.jobinfo` 

I'm guessing they're setting when the job is started and they are cleared once the job is finished.
This makes them essentially useless for the `NeomakeFinished` trigger.

I mainly use Neomake for directory makers like gulp and grunt.
All I want to know is: When the job starts and when the job ends. If there are errors or maybe warnings then of course show those.

verbose level 2 will show it starting but there's nothing for finishing. (Verbose 3 sorta says when it's finished but verbose 3 is super janky)

It would also be great to show the make id, and any other related info on its exit/finish statement. Just like the way it starts with verbose level 2.
Right now I have it just echo 'Done'.

**Comment by blueyed** [comment]
@jwmann 
Agreeded.
Error levels for msgs could be adjusted, of course. I am using 0 myself.

Need to check the jobinfo in the event.
There is also an event for finished jobs (not the whole run).

After all it boils also down to having a better statusline (more like my custom one).

**Comment by bounceme** [comment]
I think getting the `quickfixcmd` autocmds working would be better than adding new autocmds that are subtly different and specific

**Comment by bounceme** [comment]
https://github.com/tpope/vim-dispatch/commit/7bd12f84add94455931bce759f8d93354a19eea4
for reference

**Comment by jwmann** [comment]
> I think getting the `quickfixcmd` autocmds working would be better than adding new autocmds that are subtly different and specific

I don't think Neomake can rely solely on `quickfixcmd` autocmds since it works asynchronously.
A command can have one to many jobs. Being able to distinguish the difference between one Make that runs one Job and a Make that runs multiple Jobs and the start and end of those individual Makes and Jobs is pretty important.

Unless I'm misunderstanding the limitations of these autocmds, I think the custom Neomake autocmds are necessary. Of course, perhaps `NeomakeFinished` or `NeomakeJobFinished` can integrate with `quickfixcmd` but it might get messy.

**Comment by bounceme** [comment]
Dispatch fires it right before updating the quickfix window, and vim right before jumpning to first error, so presumably that is equivalent to dispatch. why can't neomake also be compatible?

**Comment by blueyed** [comment]
@bounceme 
I think we could fire it, of course - but would still have all the other hooks.
Do you need it?

btw: with https://github.com/neomake/neomake/pull/924 this might be more (ir)relevant.

**Comment by bounceme** [comment]
It isn't completely necessary, but as of now any code that uses those builtin autocmds just won't work. for example a compiler plugin, or some other plugin or something, only works with synchronous make, but if we triggered these autocmds they will be async as well

**Comment by bounceme** [comment]
For `:neomake!` is my use case

**Comment by blueyed** [comment]
@bounceme 
I see.
Should it be triggered with `NeomakeJobFinished` or `NeomakeFinished` then?
Or rather every time we change the location/quickfix list (I assume)?

**Comment by bounceme** [comment]
I tried both, neither worked for the functions in that compiler.

**Comment by blueyed** [comment]
@bounceme 
Not sure what you've tried, but you need to trigger the `QuickfixCmdPost {l,}make` for this plugin to work, of course.
So my guess is that we should do it on `NeomakeFinished`, i.e. all jobs have finished.
You could try invoking it manually where `NeomakeFinished` is invoked currently.

**Comment by bounceme** [comment]
i just bypassed the quickfixcmdpost and used `user neomakefinished` etc

**Comment by blueyed** [comment]
@bounceme 
Well, it might be good to address this properly, instead of a local workaround, if you think it makes sense.
So please consider creating a PR where `QuickfixCmdPost {l,}make` would be triggered, or at least create a separate issue to track it.

**Comment by bounceme** [comment]
Sorry, `au user neomakefinished call s:func()` didn't actually work, if that was unclear. i'll look into making a p.r. soon

**Comment by bounceme** [comment]
https://github.com/neomake/neomake/blob/master/autoload/neomake.vim#L734 is where autocmds should be triggered i think.

**Comment by bounceme** [comment]
and what would be a more accurate cmd type , cgetfile/lgetfile or make/lmake ?

**Comment by bounceme** [comment]
https://github.com/neomake/neomake/pull/958

**Comment by blueyed** [comment]
https://github.com/neomake/neomake/pull/1136 addresses this by passing in `jobinfo` in general to hooks.
